storage/secondary.go

```diff
@@ -13,25 +13,39 @@
 
 package storage
 
-import "github.com/prometheus/prometheus/pkg/labels"
+import (
+	"sync"
 
-// secondaryQuerier is a wrapper that allows a querier to be treated in best effort manner.
-// This means that a potential error on any method except Close will be passed as warning, and the result will be empty.
-// NOTE: Prometheus treats all remote storages as secondary / best effort ones.
+	"github.com/prometheus/prometheus/pkg/labels"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+)
+
+// secondaryQuerier is a wrapper that allows a querier to be treated in a best effort manner.
+// This means that an error on any method returned by Querier except Close will be returned as a warning,
+// and the result will be empty.
+//
+// Additionally, Querier ensures that if ANY SeriesSet returned by this querier's Select failed on an initial Next,
+// All other SeriesSet will be return no response as well. This ensures consistent partial response strategy, where you
+// have either full results or none from each secondary Querier.
+// NOTE: This works well only for implementations that only fail during first Next() (e.g fetch from network). If implementation fails
+// during further iterations, set will panic. If Select is invoked after first Next of any returned SeriesSet, querier will panic.
```

Comment:
Select will panic?

Reply:
Yes, what else to do if Select will be run after `once` will already complete?

Reply:
My point was s/set/Select/